paddle/fluid/framework/program_processing_test.cc

```diff
@@ -0,0 +1,221 @@
+
+  less_than_op_2->SetInput("X", {less_than_2_x->Name()});
+  less_than_op_2->SetInput("Y", {less_than_2_y->Name()});
+
+  auto* less_than_2_out = global_block->Var("Less_than_2_out");
+  less_than_2_out->SetType(proto::VarType::BOOL);
+  less_than_op_2->SetOutput("Out", {less_than_2_out->Name()});
+
+  auto* cond_op = sub_blocks[0]->AppendOp();
+  cond_op->SetType("conditional_block");
+  cond_op->SetAttr("sub_block", sub_blocks[1]);
+
+  auto* cond_x = global_block->Var("Cond_X");
+  cond_x->SetType(proto::VarType::LOD_TENSOR);
+  cond_x->SetLoDLevel(0);
+  cond_x->SetDataType(proto::VarType::FP32);
+  cond_x->SetShape({1});
+
+  cond_op->SetInput("kInputs", {cond_x->Name()});
+  cond_op->SetInput("kCondition", {less_than_2_out->Name()});
+
+  auto* cond_out = global_block->Var("Out5");
+  cond_out->SetType(proto::VarType::LOD_TENSOR);
+  cond_out->SetLoDLevel(0);
+  cond_out->SetDataType(proto::VarType::FP32);
+  cond_out->SetShape({1});
+
+  auto* scope = global_block->Var("Scope");
+  scope->SetType(proto::VarType::STEP_SCOPES);
+
+  cond_op->SetOutput("kOutputs", {cond_out->Name()});
+  cond_op->SetOutput("kScope", {scope->Name()});
+
+  auto* mul_3_x = global_block->Var("Mul_3_X");
+  mul_3_x->SetType(proto::VarType::LOD_TENSOR);
+  mul_3_x->SetLoDLevel(0);
+  mul_3_x->SetDataType(proto::VarType::FP32);
+  mul_3_x->SetShape({1000, 784});
+
+  auto* mul_3_y = global_block->Var("Mul_3_Y");
+  mul_3_y->SetType(proto::VarType::LOD_TENSOR);
+  mul_3_y->SetLoDLevel(0);
+  mul_3_y->SetDataType(proto::VarType::FP32);
+  mul_3_y->SetShape({784, 100});
+
+  auto* mul_3_out = global_block->Var("Mul_3_Out");
+  mul_3_out->SetType(proto::VarType::LOD_TENSOR);
+
+  auto* mul_op_3 = sub_blocks[1]->AppendOp();
+  mul_op_3->SetType("mul");
+  mul_op_3->SetInput("X", {mul_3_x->Name()});
+  mul_op_3->SetInput("Y", {mul_3_y->Name()});
+  mul_op_3->SetOutput("Y", {mul_3_out->Name()});
+
+  ProgramProcessor program_processor;
+  std::set<std::string> inner_inputs;
+  std::set<std::string> inner_outputs;
+
+  program_processor.GetInputsOutputsInBlock(*sub_blocks[0], &inner_inputs,
+                                            &inner_outputs);
+
+  VLOG(3) << "inner_inputs().size():" << inner_inputs.size();
+  VLOG(3) << "inner_outputs().size():" << inner_outputs.size();
+
+  ASSERT_EQ(5UL, inner_inputs.size());
+  ASSERT_EQ(4UL, inner_outputs.size());
+
+  VLOG(3) << "Before AddDependency, while op's input kX size:"
+          << while_op->Input("kX").size();
+  VLOG(3) << "Before AddDependency, while op's output kOutPuts size:"
+          << while_op->Output("kOutputs").size();
+
+  program_processor.AddDepToBlockOp(*global_block);
+  VLOG(3) << "After AddDependency, while op's input kX size:"
+          << while_op->Input("kX").size();
+  VLOG(3) << "After AddDependency, while op's output kOutPuts size:"
+          << while_op->Output("kOutputs").size();
+
+  ASSERT_EQ(8UL, while_op->Input("kX").size());
+  ASSERT_EQ(6UL, while_op->Output("kOutputs").size());
```

Comment:
Besides the size, could you also check Var name is correct?

Reply:
Tried it, but failed. I wrote the var name in the commnet.

Reply:
Done